diffrax/solver/ansr.py

```diff
@@ -0,0 +1,220 @@
+    tableau: StochasticButcherTableau
+
+    def order(self, terms):
+        return 1  # should be modified depending on tableau
+
+    def strong_order(self, terms):
+        return 0.5  # should be modified depending on tableau
+
+    def init(
+        self,
+        terms: MultiTerm[Tuple[ODETerm, ControlTerm]],
+        t0: Scalar,
+        t1: Scalar,
+        y0: PyTree,
+        args: PyTree,
+    ) -> _SolverState:
+        return None
+
+    def embed_a_lower(self, dtype):
+        """
+        Takes the lower-triangular (a_j_i) and embeds it in an n*n array.
+        """
+        num_stages = len(self.tableau.b)
+        a = self.tableau.a
+        tab_a = np.zeros((num_stages, num_stages))
+        for i, a_i in enumerate(a):
+            tab_a[i + 1, : i + 1] = a_i
+        return jnp.asarray(tab_a, dtype=dtype)
+
+    def step(
+        self,
+        terms: MultiTerm[Tuple[ODETerm, ControlTerm]],
+        t0: Scalar,
+        t1: Scalar,
+        y0: PyTree,
+        args: PyTree,
+        solver_state: _SolverState,
+        made_jump: Bool,
+    ) -> Tuple[PyTree, _ErrorEstimate, DenseInfo, _SolverState, RESULTS]:
+        del solver_state, made_jump
+
+        h = t1 - t0
+        drift, diffusion = terms.terms
+        check_diffusion_has_stla(diffusion)
+
+        # compute the Brownian increment and space-time Levy area
+        levy: LevyVal = diffusion.levy_contr(t0, t1)
+        w = levy.W
+        hh = levy.H
+        sigma = diffusion.vf(t0, y0, args)
+
+        def stage(
+            carry: tuple[int, list[PyTree]], x: tuple[jax.Array, Scalar, Scalar, Scalar]
+        ):
+            # Represents the jth stage of the SRK.
+            # carry = (j, hks_{j-1}) where
+            # hks_{j-1} = [hk1, hk2, ..., hk_{j-1}, 0, 0, ..., 0]
+            # hki = drift.vf_prod(t0 + c_i*h, y_i, args, h) (i.e. hki = h * k_i)
+            a_j, c_j, cw_j, ch_j = x
+            j, hks_j = carry
+            diffusion_control = cw_j * w + ch_j * hh
+
+            # compute Σ_{i=1}^{j-1} a_j_i hk_i
+            def lin_comb_a_j(hk_leaf):
+                return jnp.tensordot(a_j, hk_leaf, axes=1)
+
+            a_j_mult_k = jtu.tree_map(lin_comb_a_j, hks_j)
+
+            # z_j = y_0 + h (Σ_{i=1}^{j-1} a_j_i k_i) + σ * (cw_j * ΔW + ch_j * ΔH)
+            z_j = (
+                y0**ω
+                + a_j_mult_k**ω
+                + (diffusion.prod(sigma, diffusion_control)) ** ω
+            ).ω
+
+            hk_j = drift.vf_prod(t0 + c_j * h, z_j, args, h)
+            hks_j = jtu.tree_map(
+                lambda ks_leaf, k_j_leaf: ks_leaf.at[j].set(k_j_leaf), hks_j, hk_j
+            )
+            # note that carry will already contain the whole stack of
+            # k_js, so no need for second return value
+            return (j + 1, hks_j), None
+
+        a = self.embed_a_lower(jnp.dtype(y0))
+        c = jnp.insert(jnp.asarray(self.tableau.c, dtype=jnp.dtype(y0)), 0, 0.0)
```

Comment:
You can make this potentially slightly more efficient by doing the insert in `np` before casting to `jnp`. (Essentially this will force the insert to happen at compile time.)